File: src/telegram_fetcher.py
import os
import logging

import httpx

logging.basicConfig(level=logging.INFO, format="%(message)s")
log = logging.getLogger(__name__)


def fetch_telegram_posts(bot_token: str, chat_id: str, limit: int = 5) -> list[dict]:
    """
    Получает последние сообщения из Telegram-канала через Bot API.
    Не использует long polling — простой GET /getUpdates.
    """
    if not bot_token:
        log.warning("⚠️ TELEGRAM_BOT_TOKEN не задан, тестовая заглушка")
        return [{"message_id": 1, "text": "Тестовое сообщение из ТГ", "date": ""}]

    url = f"https://api.telegram.org/bot{bot_token}/getUpdates"
    params = {"timeout": 10, "allowed_updates": ["message", "channel_post"]}

    try:
        resp = httpx.get(url, params=params, timeout=15)
        data = resp.json()
        if not data.get("ok"):
            log.error(f"❌ Telegram API error: {data.get('description', 'unknown')}")
            return []

        posts = []
        seen = set()

        for update in data.get("result", []):
            msg = update.get("message") or update.get("channel_post")
            if not msg:
                continue

            text = msg.get("text") or msg.get("caption") or ""
            if not text.strip():
                continue

            mid = msg["message_id"]
            if mid in seen:
                continue
            seen.add(mid)

            posts.append({
                "message_id": mid,
                "text": text,
                "date": msg.get("date", ""),
            })

        return posts[-limit:]

    except Exception as e:
        log.error(f"❌ Ошибка Telegram: {e}")
        return []

File: src/telegram-fetcher.js
const axios = require('axios');

const TELEGRAM_BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const TELEGRAM_CHAT_ID = process.env.TELEGRAM_CHAT_ID;

/**
 * Получить последние сообщения из Telegram-канала
 * @param {number} limit - сколько последних сообщений
 * @returns {Array} массив сообщений
 */
async function fetchTelegramPosts(limit = 5) {
  if (!TELEGRAM_BOT_TOKEN) {
    console.log('⚠️ TELEGRAM_BOT_TOKEN не задан, возвращаю заглушку');
    return [
      {
        message_id: 1,
        text: 'Тестовое сообщение из Telegram о блокировках в Европе',
        date: new Date().toISOString(),
      },
    ];
  }

  try {
    const { data } = await axios.get(
      `https://api.telegram.org/bot${TELEGRAM_BOT_TOKEN}/getUpdates`,
      {
        params: { timeout: 30, offset: -limit, allowed_updates: ['message'] },
        timeout: 15000,
      }
    );

    if (!data.ok) {
      throw new Error(`Telegram API error: ${data.description || 'unknown'}`);
    }

    const posts = [];
    const seen = new Set();

    for (const update of (data.result || [])) {
      const msg = update.message || update.channel_post;
      if (!msg) continue;

      const text = msg.text || msg.caption || '';
      if (!text.trim()) continue;

      const messageId = msg.message_id;
      if (seen.has(messageId)) continue;
      seen.add(messageId);

      posts.push({
        message_id: messageId,
        text: text,
        date: new Date(msg.date * 1000).toISOString(),
        has_media: !!(msg.photo || msg.video || msg.document),
        media_group_id: msg.media_group_id,
      });
    }

    // Возвращаем последние N неповторяющихся
    return posts.slice(-limit);
  } catch (error) {
    console.error('❌ Ошибка получения постов из Telegram:', error.message);
    return [];
  }
}

/**
 * Получить ID последнего обработанного поста (из переменной)
 */
function getLastProcessedId() {
  return process.env.LAST_PROCESSED_ID ? parseInt(process.env.LAST_PROCESSED_ID) : 0;
}

module.exports = { fetchTelegramPosts, getLastProcessedId };
